File: 01-learning-spark/03-basic-sum/src/main/java/wjc/bigdata/spark/java/BasicSum.java
package wjc.bigdata.spark.java;

import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.Function2;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;

/**
 * @author: wangjunchao(王俊超)
 * @time: 2019-01-17 11:26
 **/
public class BasicSum {
    private final static Logger logger = LoggerFactory.getLogger(BasicSum.class);

    public static void main(String[] args) {
        String master = "local";
        if (args != null && args.length > 0) {
            master = args[0];
        }

        JavaSparkContext ctx = new JavaSparkContext(
                master,
                "basic-sum",
                System.getenv("SPARK_HOME"),
                System.getenv("JARS"));

        JavaRDD<Integer> rdd = ctx.parallelize(Arrays.asList(1, 2, 3, 4));

        Integer result = rdd.fold(0, (Function2<Integer, Integer, Integer>) (x, y) -> x + y);
        System.out.println(result);
        ctx.stop();
    }

}

File: spark/01-learning-spark/03-basic-sum/src/main/scala/wjc/bigdata/spark/scala/BasicSum.scala
package wjc.bigdata.spark.scala

import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD

/**
  * @author: wangjunchao(王俊超)
  * @time: 2019-01-17 13:36
  **/
object BasicSum {

    def comupteAvg(input: RDD[Int]): (Int, Int) = {
        input.aggregate((0, 0))(
            (x, y) => (x._1 + y, x._2 + 1),
            (x, y) => (x._1 + y._1, x._2 + y._2)
        )
    }

    def main(args: Array[String]): Unit = {
        val master = args.length match {
            case x: Int if x > 0 => args(0)
            case _ => "local"
        }

        val ctx = new SparkContext(master, "basic-avg",
            System.getenv("SPARK_HOME"))
        val input = ctx.parallelize(List(1, 2, 3, 4))
        val result = input.fold(0)((x, y) => x + y)
        print(result)
    }
}
